toolkit/tools/scheduler/schedulerutils/priority.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	// Finally, we want to prioritize resolving any implicit nodes in the subgraph. Implicit nodes are what blocks the
+	// scheduler from optimizing and getting a true copy of the above sub-grpah. This is done by finding a real
+	// package that should satisfy the implicit node and prioritizing that package and all of its dependencies above all
+	// else.
+	// This process is best-effort and does not guarantee that all implicit nodes will be satisfied, or that the
+	// predicted package will be the one that actually provides the implicit provide. If the prediction is incorrect
+	// the scheduler can still rely on the low-priority nodes to eventually provide the correct result.
+	logger.Log.Debugf("Setting priority for implicit provider nodes")
+	implicitNodes := subgraph.AllImplicitNodes()
+	satisfyingNodes := []*pkggraph.PkgNode{}
+	for _, implicitNode := range implicitNodes {
+		satisfyingNodesForImplicitNode, findErr := findMatchesForImplicit(implicitNode, pkgGraph)
+		if findErr != nil {
+			err = fmt.Errorf("failed to find match for implicit node (%s):\n%w", implicitNode.FriendlyName(), findErr)
+			return
+		}
+		if len(satisfyingNodesForImplicitNode) > 0 {
+			logger.Log.Debugf("For implicit (%s), we found (%v) that likely provides it.", implicitNode.FriendlyName(), satisfyingNodesForImplicitNode)
+			satisfyingNodes = append(satisfyingNodes, satisfyingNodesForImplicitNode...)
+		}
+	}
+
+	// For each real node we found, we want to prioritize it and all of its dependencies at high.
+	for _, realBuildNode := range satisfyingNodes {
+		nodePriorityMap.SetSubgraphPriority(pkgGraph, realBuildNode, HighNodePriority)
+	}
+
+	return
+}
+
+// IsElevatedPriority returns true if the given node is a medium or high priority node.
+func (s NodePriorityMap) IsElevatedPriority(node *pkggraph.PkgNode) bool {
+	pri := s.priorities[node]
+	return pri > LowNodePriority
+}
+
+// IncreaseNodePriority will increase the priority of the given node to the given priority. If the node already has a
+// higher priority, the priority will not be changed.
+func (s NodePriorityMap) IncreaseNodePriority(node *pkggraph.PkgNode, priority int) {
+	// Limit range of priority to MinPriority and MaxPriority
+	if priority < MinPriority {
+		priority = MinPriority
+	} else if priority > MaxPriority {
+		priority = MaxPriority
+	}
+
+	if s.priorities[node] < priority {
+		logger.Log.Tracef("Setting %s to priority %d", node.FriendlyName(), priority)
+		s.priorities[node] = priority
+	} else {
+		logger.Log.Tracef("Leaving %s at existing priority %d", node.FriendlyName(), s.priorities[node])
+	}
+}
+
+func (s NodePriorityMap) GetPriority(node *pkggraph.PkgNode) (priority int) {
+	return s.priorities[node]
+}
+
+// SetSubgraphPriority will increase the priority of the given node and all of its dependencies to the given priority.
+func (s NodePriorityMap) SetSubgraphPriority(pkgGraph *pkggraph.PkgGraph, node *pkggraph.PkgNode, priority int) {
+	search := traverse.BreadthFirst{}
```

Comment:
this struct has a `Visit` field that can be used over the `until` param in `seach.Walk`, it's only preference but we can get rid of the `stopSearch` value that we don't ever use

Reply:
Does read more cleanly to me as well, updated.